services/configuration/src/errors.rs

```diff
@@ -0,0 +1,114 @@
+// Copyright 2018 The Exonum Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use exonum::blockchain::{ExecutionError, StoredConfiguration};
+use exonum::crypto::Hash;
+use exonum::encoding::serialize::json::reexport::Error as JsonError;
+use exonum::helpers::Height;
+
+use transactions::Propose;
+
+/// Error codes emitted by `Propose` and/or `Vote` transactions during execution.
+#[derive(Debug)]
+#[repr(u8)]
+pub enum ErrorCode {
+    /// Next configuration is already scheduled.
+    ///
+    /// Can be emitted by `Propose` or `Vote`.
+    AlreadyScheduled = 0,
+    /// The sender of the transaction is not among the active validators.
+    ///
+    /// Can be emitted by `Propose` or `Vote`.
+    UnknownSender = 1,
+    /// The configuration in the proposal does not reference the currently active configuration.
+    ///
+    /// Can be emitted by `Propose` or `Vote`.
+    InvalidConfigRef = 2,
+    /// Current blockchain height exceeds the height of the proposal activation.
+    ///
+    /// Can be emitted by `Propose` or `Vote`.
+    ActivationInPast = 3,
+
+    /// The same configuration is already proposed.
+    ///
+    /// Specific for `Propose`.
+    AlreadyProposed = 32,
+    /// The configuration in the transaction cannot be parsed.
+    ///
+    /// Specific for `Propose`.
+    UnparseableConfig = 33,
+
+    /// The transaction references an unknown configuration.
+    ///
+    /// Specific for `Vote`.
+    UnknownConfigRef = 64,
+    /// The validator who authored the transaction has already voted for the same proposal.
+    ///
+    /// Specific for `Vote`.
+    AlreadyVoted = 65,
+}
+
+// Common error types for `Propose` and `Vote`.
+#[derive(Debug, Fail)]
+pub(crate) enum Error {
+    #[fail(display = "Next configuration is already scheduled: {:?}", _0)]
+    AlreadyScheduled(StoredConfiguration),
+
+    #[fail(display = "Not authored by a validator")]
+    UnknownSender,
```

Comment:
Should public key of the "unknown sender" will displayed?

Reply:
I assume that the error is output close to a faulty transaction (which is the case for error logging, in particular), so this information will be redundant - the sender will be displayed as a part of the transaction.